diff -up bash-5.3/builtins.h.7~ bash-5.3/builtins.h
--- bash-5.3/builtins.h.7~	2021-11-05 14:10:35.000000000 +0100
+++ bash-5.3/builtins.h	2025-07-05 17:44:03.604985262 +0200
@@ -46,6 +46,7 @@
 #define POSIX_BUILTIN	0x20	/* This builtins is special in the Posix command search order. */
 #define LOCALVAR_BUILTIN   0x40	/* This builtin creates local variables */
 #define ARRAYREF_BUILTIN 0x80	/* This builtin takes array references as arguments */
+#define REQUIRES_BUILTIN 0x100  /* rpm dependency generator */
 
 #define BASE_INDENT	4
 
diff -up bash-5.3/builtins/mkbuiltins.c.7~ bash-5.3/builtins/mkbuiltins.c
--- bash-5.3/builtins/mkbuiltins.c.7~	2025-06-25 21:49:26.000000000 +0200
+++ bash-5.3/builtins/mkbuiltins.c	2025-07-05 17:47:22.393624340 +0200
@@ -61,12 +61,17 @@ static void *xrealloc (void *, size_t);
 #define savestring(x) strcpy (xmalloc (1 + strlen (x)), (x))
 #define whitespace(c) (((c) == ' ') || ((c) == '\t'))
 
+/*  These flags are for the C code generator,
+    the C which is produced (./builtin.c)
+    includes the flags definitions found
+    in ../builtins.h */
 /* Flag values that builtins can have. */
 #define BUILTIN_FLAG_SPECIAL	0x01
 #define BUILTIN_FLAG_ASSIGNMENT 0x02
 #define BUILTIN_FLAG_LOCALVAR	0x04
 #define BUILTIN_FLAG_POSIX_BUILTIN	0x08
 #define BUILTIN_FLAG_ARRAYREF_ARG	0x10
+#define BUILTIN_FLAG_REQUIRES	0x20
 
 #define BASE_INDENT	4
 
@@ -183,13 +188,21 @@ char *arrayvar_builtins[] =
   "typeset", "unset", "wait",		/*]*/
   (char *)NULL
 };
-	
+
+/* The builtin commands that cause requirements on other files. */
+static const char *requires_builtins[] =
+{
+  ".", "command", "exec", "source", "inlib",
+  (char *)NULL
+};
+
 /* Forward declarations. */
 static int is_special_builtin (char *);
 static int is_assignment_builtin (char *);
 static int is_localvar_builtin (char *);
 static int is_posix_builtin (char *);
 static int is_arrayvar_builtin (char *);
+static int is_requires_builtin (char *);
 
 #if !defined (HAVE_RENAME)
 static int rename (char *, char *);
@@ -826,6 +839,8 @@ builtin_handler (char *self, DEF_FILE *d
     new->flags |= BUILTIN_FLAG_POSIX_BUILTIN;
   if (is_arrayvar_builtin (name))
     new->flags |= BUILTIN_FLAG_ARRAYREF_ARG;
+  if (is_requires_builtin (name))
+    new->flags |= BUILTIN_FLAG_REQUIRES;
 
   array_add ((char *)new, defs->builtins);
   building_builtin = 1;
@@ -1212,13 +1227,14 @@ write_builtins (DEF_FILE *defs, FILE *st
 		  else
 		    fprintf (structfile, "(sh_builtin_func_t *)0x0, ");
 
-		  fprintf (structfile, "%s%s%s%s%s%s, %s_doc,\n",
+		  fprintf (structfile, "%s%s%s%s%s%s%s, %s_doc,\n",
 		    "BUILTIN_ENABLED | STATIC_BUILTIN",
 		    (builtin->flags & BUILTIN_FLAG_SPECIAL) ? " | SPECIAL_BUILTIN" : "",
 		    (builtin->flags & BUILTIN_FLAG_ASSIGNMENT) ? " | ASSIGNMENT_BUILTIN" : "",
 		    (builtin->flags & BUILTIN_FLAG_LOCALVAR) ? " | LOCALVAR_BUILTIN" : "",
 		    (builtin->flags & BUILTIN_FLAG_POSIX_BUILTIN) ? " | POSIX_BUILTIN" : "",
 		    (builtin->flags & BUILTIN_FLAG_ARRAYREF_ARG) ? " | ARRAYREF_BUILTIN" : "",
+		    (builtin->flags & BUILTIN_FLAG_REQUIRES) ? " | REQUIRES_BUILTIN" : "",
 		    document_name (builtin));
 
 		  /* Don't translate short document summaries that are identical
@@ -1597,6 +1613,12 @@ is_arrayvar_builtin (char *name)
   return (_find_in_table (name, arrayvar_builtins));
 }
 
+static int
+is_requires_builtin (char *name)
+{
+  return (_find_in_table (name, requires_builtins));
+}
+
 #if !defined (HAVE_RENAME)
 static int
 rename (char *from, char *to)
diff -up bash-5.3/doc/bash.1.7~ bash-5.3/doc/bash.1
--- bash-5.3/doc/bash.1.7~	2025-04-30 15:04:32.000000000 +0200
+++ bash-5.3/doc/bash.1	2025-07-05 17:44:03.587073117 +0200
@@ -276,6 +276,14 @@ The shell becomes restricted (see
 .B "RESTRICTED SHELL"
 below).
 .TP
+.B \-\-rpm-requires
+Produce the list of files that are required for the
+shell script to run.  This implies '-n' and is subject
+to the same limitations as compile time error checking checking;
+Command substitutions, Conditional expressions and
+.BR eval
+builtin are not parsed so some dependencies may be missed.
+.TP
 .B \-\-verbose
 Equivalent to \fB\-v\fP.
 .TP
diff -up bash-5.3/doc/bashref.texi.7~ bash-5.3/doc/bashref.texi
--- bash-5.3/doc/bashref.texi.7~	2025-05-27 17:56:19.000000000 +0200
+++ bash-5.3/doc/bashref.texi	2025-07-05 17:47:59.928952167 +0200
@@ -7812,6 +7812,13 @@ standard.
 Equivalent to @option{-r}.
 Make the shell a restricted shell (@pxref{The Restricted Shell}).
 
+@item --rpm-requires
+Produce the list of files that are required for the
+shell script to run.  This implies '-n' and is subject
+to the same limitations as compile time error checking checking;
+Command substitutions, Conditional expressions and @command{eval}
+are not parsed so some dependencies may be missed.
+
 @item --verbose
 Equivalent to @option{-v}.
 Print shell input lines as they're read.
diff -up bash-5.3/eval.c.7~ bash-5.3/eval.c
--- bash-5.3/eval.c.7~	2025-03-08 00:50:03.000000000 +0100
+++ bash-5.3/eval.c	2025-07-05 17:44:03.602013353 +0200
@@ -146,7 +146,7 @@ reader_loop (void)
 
       if (read_command () == 0)
 	{
-	  if (interactive_shell == 0 && read_but_dont_execute)
+	  if (interactive_shell == 0 && (read_but_dont_execute && !rpm_requires))
 	    {
 	      set_exit_status (last_command_exit_value);
 	      dispose_command (global_command);
diff -up bash-5.3/execute_cmd.c.7~ bash-5.3/execute_cmd.c
--- bash-5.3/execute_cmd.c.7~	2025-06-05 17:02:01.000000000 +0200
+++ bash-5.3/execute_cmd.c	2025-07-05 17:49:21.016412132 +0200
@@ -608,6 +608,8 @@ async_redirect_stdin (void)
 
 #define DESCRIBE_PID(pid) do { if (interactive) describe_pid (pid); } while (0)
 
+extern int rpm_requires;
+
 /* Execute the command passed in COMMAND, perhaps doing it asynchronously.
    COMMAND is exactly what read_command () places into GLOBAL_COMMAND.
    ASYNCHRONOUS, if non-zero, says to do this command in the background.
@@ -634,6 +636,10 @@ execute_command_internal (COMMAND *comma
 
   if (breaking || continuing)
     return (last_command_exit_value);
+  if (rpm_requires && command->type == cm_function_def)
+    return last_command_exit_value =
+      execute_intern_function (command->value.Function_def->name,
+                              command->value.Function_def);
   if (read_but_dont_execute)
     return (last_command_exit_value);
   if (command == 0)
@@ -3003,7 +3009,7 @@ execute_for_command (FOR_COM *for_comman
   if (check_identifier (for_command->name, 1) == 0)
     {
       line_number = save_line_number;
-      if (posixly_correct && interactive_shell == 0)
+      if (posixly_correct && interactive_shell == 0 && rpm_requires == 0)
 	{
 	  last_command_exit_value = EX_BADUSAGE;
 	  jump_to_top_level (ERREXIT);
diff -up bash-5.3/execute_cmd.h.7~ bash-5.3/execute_cmd.h
--- bash-5.3/execute_cmd.h.7~	2024-05-06 22:02:43.000000000 +0200
+++ bash-5.3/execute_cmd.h	2025-07-05 17:44:03.603474826 +0200
@@ -22,6 +22,9 @@
 #define _EXECUTE_CMD_H_
 
 #include "stdc.h"
+#include "variables.h"
+#include "command.h"
+
 
 #if defined (ARRAY_VARS)
 struct func_array_state
diff -up bash-5.3/make_cmd.c.7~ bash-5.3/make_cmd.c
--- bash-5.3/make_cmd.c.7~	2023-12-05 16:52:24.000000000 +0100
+++ bash-5.3/make_cmd.c	2025-07-05 17:44:03.603700662 +0200
@@ -35,6 +35,8 @@
 #include "bashintl.h"
 
 #include "shell.h"
+#include "builtins.h"
+#include "builtins/common.h"
 #include "execute_cmd.h"
 #include "parser.h"
 #include "flags.h"
@@ -797,6 +799,30 @@ make_coproc_command (char *name, COMMAND
   return (make_command (cm_coproc, (SIMPLE_COM *)temp));
 }
 
+static void
+output_requirement (deptype, filename)
+const char *deptype;
+char *filename;
+{
+  static char *alphabet_set = "abcdefghijklmnopqrstuvwxyz"
+                    "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+  if (strchr(filename, '$') || (filename[0] != '/' && strchr(filename, '/')))
+    return;
+
+  /*
+      if the executable is called via variable substitution we can
+      not dermine what it is at compile time.
+
+      if the executable consists only of characters not in the
+      alphabet we do not consider it a dependency just an artifact
+      of shell parsing (ex "exec < ${infile}").
+  */
+
+  if (strpbrk(filename, alphabet_set))
+    printf ("%s(%s)\n", deptype, filename);
+}
+
 /* Reverse the word list and redirection list in the simple command
    has just been parsed.  It seems simpler to do this here the one
    time then by any other method that I can think of. */
@@ -813,6 +839,28 @@ clean_simple_command (COMMAND *command)
 	REVERSE_LIST (command->value.Simple->redirects, REDIRECT *);
     }
 
+  if (rpm_requires && command->value.Simple->words)
+    {
+      char *cmd0;
+      char *cmd1;
+      struct builtin *b;
+
+      cmd0 = command->value.Simple->words->word->word;
+      b = builtin_address_internal (cmd0, 0);
+      cmd1 = 0;
+      if (command->value.Simple->words->next)
+        cmd1 = command->value.Simple->words->next->word->word;
+
+      if (b) {
+        if ( (b->flags & REQUIRES_BUILTIN) && cmd1)
+          output_requirement ("executable", cmd1);
+      } else {
+        if (!assignment(cmd0, 0))
+          output_requirement (find_function(cmd0) ? "function" : "executable", cmd0);
+      }
+    } /*rpm_requires*/
+
+
   parser_state &= ~PST_REDIRLIST;
   return (command);
 }
diff -up bash-5.3/shell.c.7~ bash-5.3/shell.c
--- bash-5.3/shell.c.7~	2025-03-26 21:40:19.000000000 +0100
+++ bash-5.3/shell.c	2025-07-05 17:49:51.529373141 +0200
@@ -205,6 +205,9 @@ int have_devfd = 0;
 /* The name of the .(shell)rc file, DEFAULT_BASHRC is sourced by default */
 static char *bashrc_file;
 
+/* Non-zero if we are finding the scripts requirements. */
+int rpm_requires;
+
 /* Non-zero means to act more like the Bourne shell on startup. */
 static int act_like_sh;
 
@@ -272,6 +275,7 @@ static const struct {
   { "protected", Int, &protected_mode, (char **)0x0 },
 #endif
   { "rcfile", Charp, (int *)0x0, &bashrc_file },
+  { "rpm-requires", Int, &rpm_requires, (char **)0x0 },
 #if defined (RESTRICTED_SHELL)
   { "restricted", Int, &restricted, (char **)0x0 },
 #endif
@@ -510,6 +514,12 @@ main (int argc, char **argv, char **env)
     read_but_dont_execute = 1;
 #endif
 
+  if (rpm_requires)
+    {
+      read_but_dont_execute = 1;
+      initialize_shell_builtins ();
+    }
+
   if (running_setuid && privileged_mode == 0)
     disable_priv_mode ();
 
diff -up bash-5.3/shell.h.7~ bash-5.3/shell.h
--- bash-5.3/shell.h.7~	2024-10-26 19:03:04.000000000 +0200
+++ bash-5.3/shell.h	2025-07-05 17:50:11.692469721 +0200
@@ -105,6 +105,7 @@ extern int startup_state;
 extern int reading_shell_script;
 extern int ssh_reading_startup_files;
 extern int shell_initialized;
+extern int rpm_requires;
 extern int bash_argv_initialized;
 extern int subshell_environment;
 extern int current_command_number;
