diff -up bash-5.3/Makefile.in.9~ bash-5.3/Makefile.in
--- bash-5.3/Makefile.in.9~	2025-05-16 20:56:31.000000000 +0200
+++ bash-5.3/Makefile.in	2025-07-05 17:55:26.458814576 +0200
@@ -282,7 +282,7 @@ RL_INCLUDEDIR = @RL_INCLUDEDIR@
 
 READLINE_LIB = @READLINE_LIB@
 READLINE_LIBRARY = $(RL_LIBDIR)/libreadline.a
-READLINE_LDFLAGS = -L${RL_LIBDIR}
+#READLINE_LDFLAGS = -L${RL_LIBDIR}
 READLINE_DEP = @READLINE_DEP@
 
 # The source, object and documentation of the GNU Readline library.
@@ -334,7 +334,7 @@ HIST_ABSSRC = ${topdir}/$(HIST_LIBDIR)
 
 HISTORY_LIB = @HISTORY_LIB@
 HISTORY_LIBRARY = $(HIST_LIBDIR)/libhistory.a
-HISTORY_LDFLAGS = -L$(HIST_LIBDIR)
+#HISTORY_LDFLAGS = -L$(HIST_LIBDIR)
 HISTORY_DEP = @HISTORY_DEP@
 
 HIST_INCLUDEDIR = @HIST_INCLUDEDIR@
@@ -355,7 +355,7 @@ TERM_ABSSRC = ${topdir}/$(TERM_LIBDIR)
 
 TERMCAP_LIB = @TERMCAP_LIB@
 TERMCAP_LIBRARY = $(TERM_LIBDIR)/libtermcap.a
-TERMCAP_LDFLAGS = -L$(TERM_LIBDIR)
+#TERMCAP_LDFLAGS = -L$(TERM_LIBDIR)
 TERMCAP_DEP = @TERMCAP_DEP@
 
 TERMCAP_SOURCE	= $(TERM_LIBSRC)/termcap.c $(TERM_LIBSRC)/tparam.c
diff -up bash-5.3/aclocal.m4.9~ bash-5.3/aclocal.m4
--- bash-5.3/aclocal.m4.9~	2025-05-16 18:40:49.000000000 +0200
+++ bash-5.3/aclocal.m4	2025-07-05 17:54:52.765250287 +0200
@@ -1846,8 +1846,8 @@ eval ac_cv_rl_includedir=${ac_cv_rl_pref
 eval ac_cv_rl_libdir=${ac_cv_rl_prefix}/lib
 
 LIBS="$LIBS -lreadline ${TERMCAP_LIB}"
-CFLAGS="$CFLAGS -I${ac_cv_rl_includedir}"
-LDFLAGS="$LDFLAGS -L${ac_cv_rl_libdir}"
+[ "${ac_cv_rl_prefix}" != "/usr" ] && CFLAGS="$CFLAGS -I${ac_cv_rl_includedir}"
+[ "${ac_cv_rl_prefix}" != "/usr" ] && LDFLAGS="$LDFLAGS -L${ac_cv_rl_libdir}"
 
 AC_CACHE_VAL(ac_cv_rl_version,
 [AC_RUN_IFELSE([AC_LANG_SOURCE([[
