diff -up bash-5.2-rc1/parse.y.extended_quote bash-5.2-rc1/parse.y
--- bash-5.2-rc1/parse.y.extended_quote	2022-06-19 03:07:53.037188368 +0200
+++ bash-5.2-rc1/parse.y	2022-06-19 03:10:24.689099047 +0200
@@ -3835,7 +3835,7 @@ parse_matched_pair (qc, open, close, len
 		      nestlen = strlen (nestret);
 		    }
 #endif
-		  else if ((rflags & P_DQUOTE) == 0)
+		  else if (extended_quote || (rflags & P_DQUOTE) == 0)
 		    {
 		      nestret = sh_single_quote (ttrans);
 		      free (ttrans);
